### ISSUE TITLE
Improve Zod Type Inspection

### ISSUE THREAD

**Issue by rubber-duck-software** [colinhacks/zod#1833]
Currently, there is no general purpose way to retrieve the types contained within a zod type. There a disparate apis such a `.shape` for ZodObject, `.element` for ZodArray, `.items` for ZodTuple, `.unwrap` for ZodNullable, and `innerType` for ZodEffects. As the Zod ecosystem grows, more and more third party libraries want to inspect zod types. 

I propose a new api on ZodType which enables general purpose inspection. A working title for this might be `innerTypes`.

Example of usage 
```ts
z.object({
    foo: z.string(),
    bar: z.number(),
    nested: z.object({
        baz: z.boolean()
    })
}).innerTypes() // returns [{ path: 'foo', schema: ZodString }, { path: 'bar', schema: ZodNumber }, { path: 'nested', schema: ZodObject }]
```

**Comment by rubber-duck-software** [comment]
I am happy to create a PR for this, there is a point of ambiguity which should be addressed. What is the path for types like ZodArray and ZodRecord? My proposal would be using a convention like `*` to represent a wildcard path.

**Comment by colinhacks** [comment]
I think this would best be addressed with a visitor pattern that traverses the entire schema hierarchy. This also probably shouldn't be implemented as a method since it's a lot of code that will rarely be used. Something like this:

```ts
import {walk} from "zod"

const mySchema = /* any zod schema */;

walk(mySchema, (ctx)=>{
  ctx.schema; // z.ZodFirstPartySchemaTypes
  if(schema._def.type === "
})
```

This would do a depth-first search. It would have to terminate at ZodLazy to avoid infinite loops. 

As you alluded to, this will be useful for a lot of libraries in the Zod ecosystem, so we could consider branding this as some sort of "generator" API.

**Comment by rubber-duck-software** [comment]
Using a visitor pattern might be more practical since you're right that it will rarely be used in user land. The limitation with the visitor pattern however is that I can't see a way for a ZodType to declare its own innerTypes. Meaning if I as a library create a new class extending ZodType, its innerTypes won't be found by the `walk` function. But perhaps you never intended to have people extend ZodType themselves. 

Whatever the API for this looks like, I feel like it should enable answering questions like the following
1. What is the ZodSchema at `foo.bar`?
2. Underneath any Effects, Nullable, Default etc. what is the type of this ZodSchema?
3. Is this ZodSchema some sort of Number Array? 

These are the questions which I think would be useful to be able to answer in a first class way, but there are probably others. 

Would the `walk` function be able to expose "where" in the ZodType this schema is?

**Comment by colinhacks** [comment]
> These are the questions which I think would be useful to be able to answer in a first class way, but there are probably others.

I agree with you completely and this is my biggest issue with Zod's current architecture. Doing these things are sort of possible but it's never going to be easy to answer these questions, or even ask them. Some of them aren't even well-defined.

> What is the type of this ZodSchema? Is this ZodSchema some sort of Number Array?
Does `(number | undefefined)[]` count? What if its a union and one of the elements is `z.number().array()`? Input or output type? Remember that ZodEffects contain transforms. Zod has no way of knowing what the type is at the output of a transform. And now with ZodPipeline, you can now connect together a bunch of schemas. Does it count as a "number array" if it's an array of numbers at any point along the pipeline? It goes on and on. This complexity is why I'm so averse to things like `z.discriminatedUnion` which essentially require a (hacky) `walk` implementation to work properly. So many edge cases.

I'm likely doing to publish a `zod-lite` library with a dramatically different architecture and massively reduced scope that will address your use cases better. I have some early sketches.

**Comment by rubber-duck-software** [comment]
Can we evaluate the proposed `innerTypes()` method as another possible solution to this problem? I think it is powerful enough to answer all these questions, or at least answer them as precisely as they can be answered in Zod. (Also I'm excited to hear about a possible `zod-lite` package - I think that would be really well received)

But, for now, how could these questions be answered in Zod? With the proposed `innerTypes()` API, I believe it might look something like below.

### What is the ZodSchema at foo.bar?
```js
function getAtPathSegment(schema: ZodTypeAny|undefined, pathSegment:string) {
    if(!schema) return undefined
    const match = schema.innerTypes().find(t => t.path === pathSegment)
    if (match) return match
    else return getAtPathSegment(schema.innerTypes().find(t => t.path === ''), pathSegment)
}

function getAtPath(schema: ZodTypeAny, path: string) {
    const segments = path.split('.')
    let cur = schema
    for (let p of segments) {
        cur = getAtPathSegment(cur, p)
    }
    return cur
}
```

### Underneath any Effects, Nullable, Default etc. what is the type of this ZodSchema?

Technically, this has to be an array since ZodUnion allows a type to be unwrapped in multiple possible ways
```js
function unwrap(schema: ZodTypeAny) {
    const inner = schema.innerTypes()
    const noPath = inner.filter(t => t.path === '').map(t => t.schema)
    if (noPath.length) {
        return noPath.flatMap(schema => unwrap(schema))
    } else {
        return [schema] // Unwrap stops here
    }
}
```

### Is this ZodSchema some sort of Number Array?
Let's say we're not talking about a typescript number array (I agree, this would be impossible to detect), but we're looking for the Zod Type of Number Array. But the ZodArray might be wrapped in some ways and the ZodNumber might be wrapped in some ways. This question seems answerable. 

```js
function isNumberArraySchema(schema: ZodTypeAny) {
    const outers = unwrap(schema) // see function definition above
    const arrays = outers.filter(o => o instanceof ZodArray)
    return arrays.some(a => unwrap(a.element).some(e => e instanceof ZodNumber))
}
```

Obviously these examples are pretty contrived and these function implementations are untested, but I think `innerTypes` is a powerful enough API to solve some of the most common problems Zod libraries encounter. 

Additional comments: Assuming you're on board with this API, there would still have to be discussions about how this behaves with ZodUnion, ZodIntersection, ZodPipeline, etc. Some of these edge cases will still be a bit messy, but I think there is a "good enough" answer for all these edge cases such that we should let what's perfect get in the way of what's good. If the API doesn't flawlessly deal with the complexities of input/output types, we can make a well-documented decision on what the API will do and still deliver tons of value to package authors.

**Comment by colinhacks** [comment]
zI do see the appeal here. I wish I'd thought ahead enough to provide a consistent interface `inner: { key: string | null; schema: ZodType }[]` on the base class for extracting it's internal schemas.

I don't quite see why ZodUnion and ZodIntersection would be messy - they'd just return their components, no?

ZodPipeline and ZodFunction are fuzzier. Perhaps the `key` could be tagged as `"input"/"
args"` and `"output"/"returnType"`? Bit weird. 

My issue with this is that this is a very advanced use case, and you can write this function using exactly the behavior you need with an `innerTypes` function. This avoids any of the fuzziness/arbitrariness in how Zod defines an "inner schema" and is the easiest part of this whole thing. Writing the code samples you provided is more complex than writing the function to extract the inner schemas, and you can do proper error handling if you encounter an instance of a subclass you can't or don't want to handle (e.g. ZodFunction)

```ts
import { z } from "./src";

function innerTypes(schema: z.ZodTypeAny) {
  if (schema instanceof z.ZodObject) {
    return Object.entries(schema.shape);
  } else {
    return [];
  }
}
```

**Comment by rubber-duck-software** [comment]
The reason to makes this function a part of zod instead of leaving it in user land is that Zod is subclassable. I can't find a way for a package to implement any of the functions above given that the user might have their own subclasses. This one of the problems you wrote about it in #2106. 

So, complex cases. ZodUnion, ZodIntersection, ZodFunction, ZodPipeline

I think ZodUnion should return an inner type for each option in the union. The path for each option should be the empty string. 

The more I think about it, I think ZodIntersection should return an empty array for innerTypes. This type does not contain other types, it is not composing. Rather it fused together two types in a way that can't be separated cleanly. (Just an opinion)

I think ZodFunction should also have no innerTypes. You cannot access the parameter type or return type via JSON path, therefore I think it is reasonable say this have no innerTypes. 

Finally ZodPipeline. Admittedly I'm not super familiar with this one as I don't often use it in my projects, but I'd expect using the output type would be reasonable. 


Again- I don't think there's a perfect answer for these cases, but I think there's a good enough answer such that this API can provide a lot a value.
